src/async_fuse/memfs/cache/storage_manager.rs

```diff
@@ -0,0 +1,775 @@
+            })
+            .collect()
+    }
+
+    /// Store data into storage.
+    pub async fn store(
+        &self,
+        ino: INum,
+        offset: usize,
+        data: &[u8],
+        mtime: SystemTime,
+    ) -> SystemTime {
+        // Check if the cache is valid.
+        let invalid = {
+            let guard = pin();
+            let cache_mtime = self.mtimes.get(&ino, &guard);
+            cache_mtime != Some(&mtime)
+        };
+
+        if invalid {
+            self.storage.invalidate(ino).await;
+        }
+
+        if data.is_empty() {
+            // The cache is invalid, but no new blocks will be loaded into the cache
+            // thus the mtime of this file is removed.
+            if invalid {
+                self.mtimes.remove(&ino);
+            }
+            // No data will be written, so the passed-in mtime will be passed-out changelessly.
+            return mtime;
+        }
+
+        let start_block = self.offset_to_block_id(offset);
+
+        let io_blocks = self.make_io_blocks_from_slice(offset, data);
+
+        self.store_blocks(ino, start_block, io_blocks).await;
+
+        // As the cache is overwritten, the cache mtime should be set to now.
+        let new_mtime = SystemTime::now();
+        self.mtimes.insert(ino, new_mtime);
+
+        new_mtime
+    }
+
+    /// Remove a file from the storage.
+    pub async fn remove(&self, ino: INum) {
+        self.mtimes.remove(&ino);
+        self.storage.remove(ino).await;
+    }
+
+    /// Flush the cache to the persistent layer.
+    pub async fn flush(&self, ino: INum) {
+        self.storage.flush(ino).await;
+    }
+
+    /// Flush all file in the cache to persistent layer.
+    pub async fn flush_all(&self) {
+        self.storage.flush_all().await;
+    }
+
+    /// Truncate a file, from size of `from` to size of `to`.
+    /// Both `from` and `to` are in bytes, and are excluded from the range.
+    /// After the truncating, the valid range of a file in the storage
+    /// is `[0, to)`.
+    ///
+    /// This method performs a store operation, therefore it also takes a `mtime` and returns a new `mtime` like `store` method.
+    pub async fn truncate(
+        &self,
+        ino: INum,
+        from: usize,
+        to: usize,
+        mtime: SystemTime,
+    ) -> SystemTime {
+        if from <= to {
+            return mtime;
+        }
+
+        let from_block = self
+            .offset_to_block_id(from.overflow_sub(1))
+            .overflow_add(1);
+        let to_block = self.offset_to_block_id(to.overflow_sub(1)).overflow_add(1);
+
+        self.storage.truncate(ino, from_block, to_block).await;
+
+        // fill the rest bytes with zeros.
+        let fill_end = to_block.overflow_mul(self.block_size);
+        let fill_start = to;
+        let fill_len = fill_end.overflow_sub(fill_start);
+        let fill_content = vec![0_u8; fill_len];
+        self.store(ino, fill_start, &fill_content, mtime).await
```

Comment:
We can do this in `Storage::truncate`, in order to handle the truncating in a single function call.

Reply:
There is no need to handle the truncating in a single function call, although this causes multiple calls on disk cache.

There should not be one-to-one correspondence between function call and disk access in the disk cache, the cache has the responsibility to optimize the IO operation.